Sources/AWSLambdaRuntimeCore/LambdaContext.swift

```diff
@@ -49,40 +50,132 @@ extension Lambda {
 extension Lambda {
     /// Lambda runtime context.
     /// The Lambda runtime generates and passes the `Context` to the Lambda handler as an argument.
-    public final class Context: CustomDebugStringConvertible {
+    public struct Context: BaggageContext.Context, CustomDebugStringConvertible {
+        /// Used to store all contents of the context and implement CoW semantics for it.
+        private var storage: Storage
+
+        final class Storage {
+            var baggage: Baggage
+
+            let invokedFunctionARN: String
+            let deadline: DispatchWallTime
+            let cognitoIdentity: String?
+            let clientContext: String?
+
+            // Implementation note: This logger is the "user provided logger" that we will log to when `context.logger` is used.
+            // It must be updated with the latest metadata whenever the `baggage` changes.
+            var _logger: Logger
+
+            let eventLoop: EventLoop
+            let allocator: ByteBufferAllocator
+
+            init(
+                baggage: Baggage,
+                invokedFunctionARN: String,
+                deadline: DispatchWallTime,
+                cognitoIdentity: String?,
+                clientContext: String?,
+                logger: Logger,
+                eventLoop: EventLoop,
+                allocator: ByteBufferAllocator
+            ) {
+                self.baggage = baggage
+                self.invokedFunctionARN = invokedFunctionARN
+                self.deadline = deadline
+                self.cognitoIdentity = cognitoIdentity
+                self.clientContext = clientContext
+                self._logger = logger
+                self.eventLoop = eventLoop
+                self.allocator = allocator
+            }
+        }
+
+        /// Contains contextual metadata such as request and trace identifiers, along with other information which may
+        /// be carried throughout asynchronous and cross-node boundaries (e.g. through HTTPClient calls).
+        public var baggage: Baggage {
+            get {
+                self.storage.baggage
+            }
+            set {
+                if isKnownUniquelyReferenced(&self.storage) {
```

Comment:
please correct me if i am wrong, as far as I understand it:

1. everything (new storage object) is copied every time baggage is updated, which will happen every time new span is created = often

2. this does implement copy on write however changing of the storage is not thread safe which I think is not a nice API for user who would need to sync the changes himself

see discussion here https://gist.github.com/drewmccormack/b1c4487935cf3c3e0a5feaf488a95ebd#gistcomment-2826755

Reply:
Eh seems I have gotten it wrong here, could be.

With the latest changes in context perhaps actually we can get away without the Storage completely after all actually. I'll revisit shortly

Reply:
@pokryfka, so no. This is implementation correct. (You had me worried there for a second! 😉)

This is as thread safe as an `Int` or any other value semantics type is.

`isKnownUniquelyReferenced` is commonly misunderstood, so I understand where your confusion is coming from. However such implementation is safe by itself. Now how _someone_ uses this type may or may not be thread-safe -- and that is what the example you're linking is doing wrong. The same kind of race can be exhibited on even a plain integer value. This type provides value semantics guarantees, and no more.

--- 

> everything (new storage object) is copied every time baggage is updated, which will happen every time new span is created = often

I disagree about the "often" quantification used here. It is _by far_ not as often as a raw struct would be copying -- each passing around and reassigning to a new variable even would technically then be copying.

We're getting the best of both worlds; and simply following the known CoW semantics that are frequent in Swift.

> this does implement copy on write however changing of the storage is not thread safe which I think is not a nice API for user who would need to sync the changes himself

The operations in this piece of code _are_ safe. If they weren't then Array and all of Swift's collections are screwed as well 😉 We are providing the same semantics as those types.

Reply:
> > everything (new storage object) is copied every time baggage is updated, which will happen every time new span is created = often
> 
> 
> 
> I disagree about the "often" quantification used here. It is _by far_ not as often as a raw struct would be copying -- each passing around and reassigning to a new variable even would technically then be copying.
> 
> 
> 
> We're getting the best of both worlds; and simply following the known CoW semantics that are frequent in Swift.
> 
> 

it used to be a "const reference" and so only the reference (pointer) was copied.

I dont know how expensive is copying of a small struct, and how much more expensive is allocating the space for it on heap rather than on stack; perhaps its not a real issue at all.

if the goal was to avoid copying of the lambda context, perhaps `Context` should have a reference to "static storage" and a reference to CoW "mutable storage" (the baggage). That way 2 references would always be copied but the content of the "static storage" would not need to be copied every time the baggage changes.


> 
> > this does implement copy on write however changing of the storage is not thread safe which I think is not a nice API for user who would need to sync the changes himself
> 
> 
>  
> The operations in this piece of code _are_ safe. If they weren't then Array and all of Swift's collections are screwed as well 😉 We are providing the same semantics as those types.

Thank you for rechecking it.
I do find it a bit confusing...

if I understand that it will be safe because `set` is `mutating`.

```swift
struct Context {

func notSafeUpdateBaggage(_ baggage: Baggage) {
 // update storage
}

// need to make a copy which will increase the storage reference counter
mutating func updateBaggage(_ baggage: Baggage) {
 // update storage
}

}
```

Reply:
> if the goal was to avoid copying of the lambda context, perhaps Context should have a reference to "static storage" and a reference to CoW "mutable storage" (the baggage). That way 2 references would always be copied but the content of the "static storage" would not need to be copied every time the baggage changes.

Yeah if it becomes a problem I think we could do that 🤔 👍 

> I do find it a bit confusing...
> 
> if I understand that it will be safe because set is mutating.

Maybe this example can illustrate: Because this is a struct, in order to change it the function must be mutating (or `set {}` which implicitly is).

Then if anyone has: 

```
let context = ...
```

they can't mutate it unless a second reference is created:

```
let context = ...
var changeMe = context
```

which ensures that we'll have the +1 count there and will copy on mutation.

It *still* is possible to make bugs like:

```
class Me { 
var onlyReference: Lambda.Context = ... 
for t in threads { 
  t.run { onlyReference.mutate() }
}
```

that remains illegal as usual, but not because of the CoW or Context type being wrong but how the user _uses_ it being unsafe. The same example is true for e.g. `var int: Int` which you'd try to mutate `+=1` from many threads. There's nothing wrong about the Int type, but about the unsafe modifications to where we store it.